### PR TITLE
Improve plugins admin

### PR DESCRIPTION
**Pull request by trasher**

### CODE REVIEW

**Comment by AnaelMobilia** [comment]
This could fix #1369 and fix #989 (you should known this ticket ;-)) ?

**Comment by trasher** [comment]
Indeed; thanks for the pointer! :D

**Comment by orthagh** [comment]
seems good.

My 2 cents:

- The current gui doesn't match other pages:
image
image

- we have an "hidden" feature for massive checboxes, you can shift/ctrl select them. It doesn't work in the plugin page. It's minor, i think the code about that is defined in Search class (and maybe it should be in a js file)

- my idea about plugin page, is that button (or line) should be post/transformed by ajax. single installation/activation/etc will not reload entire page.

**Comment by trasher** [comment]
Using global massive actions with the plugins (and rework to provide same features as others - as stated in #989) seems a very huge work.

I'll drop this part from the current PR; the goal was to quick improve the page, not rewrite it entirely.

File: inc/plugin.class.php [hunk 1]
```diff
@@ -500,26 +500,31 @@ function listPlugins() {
...
                      }
                      if ($plug['state'] == self::NOTUPDATED) {
                         $msg = _x('button', 'Upgrade');
                      } else {
                         $msg = _x('button', 'Install');
                      }
                      if ($do_install) {
-                        Html::showSimpleForm(static::getFormURL(), array('action' => 'install'),
+                        $out .= Html::getSimpleForm(static::getFormURL(), array('action' => 'install'),
                                              $msg, array('id' => $ID));
```
Comment (orthagh) [inline inc/plugin.class.php]: re-indent here

File: inc/plugin.class.php [hunk 2]
```diff
@@ -638,44 +655,85 @@ function uninstall($ID) {
             $_SESSION['glpi_plugins']['temp'] = $this->fields['directory']; // For autoloader
             $function();
             unset($_SESSION['glpi_plugins']['temp']);
+         } else {
+            Session::addMessageAfterRedirect(
+               sprintf(__('Plugin %1$s has no install function!'), $this->fields['name']),
```
Comment (orthagh) [inline inc/plugin.class.php]: uninstall ?

File: inc/plugin.class.php [hunk 3]
```diff
@@ -1311,4 +1311,39 @@ static function haveImport() {
...
+    * @param string $typ Either 'php' or 'core', defaults to 'core'
+    * @param string $min Minimal required version
+    * @param string $max Maximal required version
+    *
+    * @since 9.2
+    *
+    * @return string
+    */
+   static public function messageIncompatible($type = 'core', $min = null, $max = null) {
+      $type = ($type === 'core' ? __('GLPi') : __('PHP'));
```
Comment (remicollet) [inline inc/plugin.class.php]: GLPI ;)

File: inc/plugin.class.php [hunk 4]
```diff
@@ -500,25 +500,31 @@ function listPlugins() {
...
 
                      $function   = 'plugin_' . $plug['directory'] . '_check_prerequisites';
-                     $do_install = true;
                      if (function_exists($function)) {
+                        ob_start();
                         $do_install = $function();
+                        if (!$do_install) {
+                           $out .= '<span class="error">' . ob_get_contents() . '</span>';
+                        }
+                        ob_end_clean();
```
Comment (remicollet) [inline inc/plugin.class.php]: what happen is the plugin display something which is not a error, e.g; progress information ?

Comment (trasher) [inline inc/plugin.class.php]: This catch the check_prerequisites; not the install, I don't think plugins do progress infos checking PHP and/or glpi version, do they?

Comment (remicollet) [inline inc/plugin.class.php]: so forget   (btw, I don't really like the ob_* way...)

Comment (trasher) [inline inc/plugin.class.php]: I do not like that at all... But since plugin (and core also) used to "echo" directly, there is no real alternative :'(

**Pull request merged**